### ISSUE TITLE
Improvements to debug_nans

### ISSUE THREAD

**Issue by emilyfertig** [jax-ml/jax#25643]
This issue tracks improvements to debug_nans/debug_infs. See the description of #25519 for some examples of status quo behavior.

- [ ] NaNs in the forward pass of a shard_map function currently only report "Invalid value in a sharded computation," and the line where the shard_map function was called. The traceback should extend to the lax/jnp op within the shard_map function that produced the NaN.
- [ ] Avoid re-running functions with side effects and collectives.
- [ ] Consider stopping the traceback at the `jax.numpy` boundary instead of the lax primitive that NaNed in a jnp function, or using the `JAX_TRACEBACK_FILTERING` flag to switch this behavior.
- [ ] Improve the error messages for NaNs in the backward pass. Currently they report a stacktrace of line numbers, but it would be nice to have a regular Python stacktrace with code highlighted.
- [ ] For pmap, the Python dispatch path reports "Invalid value in parallel computation" and the traceback stops at the call to the pmap function. It should extend to the arithmetic op where the NaN occurred, like it does for the  C++ dispatch path.

**Comment by Stella-S-Yan** [comment]
> * For pmap, the Python dispatch path reports "Invalid value in parallel computation" and the traceback stops at the call to the pmap function. It should extend to the arithmetic op where the NaN occurred, like it does for the  C++ dispatch path.

Since `pmap` is deprecated and `shard_map` is the current alternative, I examined the dispatch path of shard_map. Both fast and slow dispatch correctly output errors pointing to the relevant line of code, and there is no occurrence of the "Invalid value in parallel computation" message.

Here is the test case:
```
import jax
import jax.numpy as jnp
from jax.experimental.shard_map import shard_map

jax.clear_caches()
jax.config.update("jax_log_compiles", False)

P = jax.sharding.PartitionSpec
mesh = jax.make_mesh((1,), ('x',))

@jax.jit
def f(x):
    y = jnp.square(x)
    return jnp.log(-y)

f_shard_map = shard_map(f, mesh=mesh, in_specs=(P('x'),), out_specs=P('x'))

with jax.debug_nans(True):
    one = jnp.ones([1])
   
    print(f_shard_map(jnp.zeros([1]))) 
    print(f_shard_map(one))
```
Slow path output is : 
```
with jax.debug_nans(True):
  one = jnp.ones([1])

  # print(f_shard_map(jnp.zeros([1]))) 
  print(f_shard_map(one))
```
```
Traceback (most recent call last):
  File "/home/stellasyan/jax/tests/syan_test2.py", line 24, in <module>
    print(f_shard_map(one)) 
          ^^^^^^^^^^^^^^^^
FloatingPointError: Invalid value (nan) encountered in sharded computation.
--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```
And fast path output is:
```
with jax.debug_nans(True):
  one = jnp.ones([1])

  print(f_shard_map(jnp.zeros([1]))) 
  print(f_shard_map(one)) 
```

```
[-inf]
Traceback (most recent call last):
  File "/home/stellasyan/jax/tests/syan_test2.py", line 24, in <module>
    print(f_shard_map(one)) 
          ^^^^^^^^^^^^^^^^
FloatingPointError: Invalid value (nan) encountered in sharded computation.
--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

Both the slow and fast paths produce the same error message. However, I assume we would prefer the slow path to provide more detailed information by pointing directly to the log operation.
